src/module_3/Module3_Milestone1.md

```diff
@@ -0,0 +1,1825 @@
+
+Finally, comparing the model’s performance on training and validation datasets reveals similar error rates, suggesting effective generalization to new data. This consistency confirms that the model is not overfitting; it is not merely memorizing the training data or modelling its noise, but is learning to generalize from it.
+
+#### 3.2 Lasso Regression
+
+
+
+```python
+# Train the Logistic Regression model
+
+# Prepare the plot environments for both training and validation
+fig_train, axes_train = plt.subplots(1, 2, figsize=(14, 6))
+fig_train.suptitle('Training Data - Performance Curves Across Different C Values')
+fig_val, axes_val = plt.subplots(1, 2, figsize=(14, 6))
+fig_val.suptitle('Validation Data - Performance Curves Across Different C Values')
+
+
+# Define hyperparameter for the level of regularization
+cs = [1e-8, 1e-6, 1e-4, 1e-2, 1, 100, 1e4, None]
+
+
+for c in cs:
+    if c == None:
+        lr = LogisticRegression(penalty = None)
+    else:
+        lr = LogisticRegression(penalty='l1', C=c, solver="saga")
+    
+    # Fit the model on scaled training data
+    lr.fit(X_train_scaled, Y_train)
+
+    # Predict probabilities for the positive class on the training set
+    train_proba = lr.predict_proba(X_train_scaled)[:, 1]
+    plot_curves_with_hyperparameter(axes_train, Y_train, train_proba, f'Train C={c}', curve_type="both")
+
+
+    # Predict probabilities for the positive class on the validation set
+    val_proba = lr.predict_proba(X_val_scaled)[:, 1]
+    plot_curves_with_hyperparameter(axes_val, Y_val, val_proba, f'Val C={c}', curve_type="both")
+
+# Add legends and show plots
+axes_train[0].legend(loc='upper right')
+axes_train[1].legend(loc='lower right')
+axes_val[0].legend(loc='upper right')
+axes_val[1].legend(loc='lower right')
+
+plt.tight_layout()
+fig_train.show()
+fig_val.show()
+```
+
+    /tmp/ipykernel_82778/1816860315.py:39: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
+      fig_train.show()
+    /tmp/ipykernel_82778/1816860315.py:40: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
+      fig_val.show()
+
+
+
+    
+![png](Module3_Milestone1_files/Module3_Milestone1_39_1.png)
+    
+
+
+
+    
+![png](Module3_Milestone1_files/Module3_Milestone1_39_2.png)
+    
+
+
+This time, with Lasso regression, we can observe that for the Precision-Recall plot indicates that the model works best for $c=10^{-8}$ and $c=10^{-6}$ the model performs the best, as it contais a higher AUC. Nevertheless, when looking at the ROC curve, we can observe that these values provide the lowest AUC. Therefore, the level of regularization does not seem to make a great impact. 
+
+Additionally, we can observe that the performance is similar to Ridge.
+
+Disclaimer: it is worth mentioning that the Precision-Recall curve for $c=10^{-6}$ does not seem correct, as it does not match with its corresponding AUC.
```

Comment:
What happens here is that you have a lot of regularization, probably all weights are 0. So you either classify everything as 0 or 1. That is why you get a 45 line (of 2 points) in the ROC and the PR curve you get